Bare bones UIToolkit version of the input asset editor

andrew-oc opened the pull request:
### Description

This is mostly a base to build new input asset functionality using UIToolkit.

andrew-oc committed:
Basic UITK version of input asset editor

Mostly read only for now. Doesn't contain things like drag and drop for list views, adding/delete/editing action maps, action, or bindings.

andrew-oc committed:
Implemented a view architecture for views to only update when relevant view state has changed

* Fixed some styling across light and dark themes, and other styling issues
* Show correct icons on actions and bindings

andrew-oc committed:
Fix InputControlPicker changes not being saved

andrew-oc committed:
Removing meta file that should have been removed before

andrew-oc committed:
State is now closer to properly immutable and removed usages of ReactiveProperty

* Removed ReactiveProperty properties from state. All changes now go through the state container, and that should be the only way to get state change events.
* Added the With method to GlobalInputActionsEditorState to make it easier to mutate into new instances.
* Mutating state actions now return new instance of the state and the state container ensures that the latest version is used everywhere.

andrew-oc committed:
Remove Global from all names and moved everything to UITKAssetEditor folder

andrew-oc committed:
Add control scheme editor

Also renamed UIToolkitView to ViewBase

andrew-oc committed:
Remove temporary test assets

andrew-oc committed:
Removed test editor window and made the new editor the default for opening input assets

andrew-oc committed:
Change yamato config to build develop-v2 branch, but only build against trunk

ekcoh commented on Assets/Samples/CustomComposite/CustomComposite.cs:
```diff
@@ -175,6 +177,27 @@ public override void OnGUI()
+		    showInputField = true
+	    };
+
+	    // Note: For UIToolkit sliders, as of Feb 2022, we can't register for the mouse up event directly
```
Run formatter?

andrew-oc replied:
Done

ekcoh commented on Assets/Samples/CustomComposite/CustomComposite.cs:
```diff
@@ -175,6 +177,27 @@ public override void OnGUI()
+	    };
+
+	    // Note: For UIToolkit sliders, as of Feb 2022, we can't register for the mouse up event directly
+        // on the slider because an element inside the slider captures the event. The workaround is to
```
This sounds similar to how things usually work, e.g. the thumb grabs the context/focus, why are we even listening to mouse event here? E.g. any kind of input could be used to set the value right? Aren't there a "changed event" or similar signal on the slider mapping to the value and not trigger of the value (the classic UI way)?

andrew-oc replied:
The same bug that makes the thumb not send events also breaks the slider entirely if you use RegisterValueChangedCallback, so we can't go that way. This comment did cause me to find a bug in the composite binding view though, so nice one :)

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionEditorWindow.cs:
```diff
@@ -34,7 +34,7 @@ internal class InputActionEditorWindow : EditorWindow, IDisposable
         /// </summary>
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId = "line", Justification = "line parameter required by OnOpenAsset attribute")]
-        [OnOpenAsset]
+        // [OnOpenAsset]
```
Delete or temporary thing that was forgotten about?

andrew-oc replied:
Deleted

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/ParameterListView.cs:
```diff
@@ -197,6 +198,24 @@ public void Initialize(Type registeredType, ReadOnlyArray<NamedValue> existingPa
 
                 m_ParameterEditor = (InputParameterEditor)Activator.CreateInstance(parameterEditorType);
+
+                // We have to jump through some hoops here to create instances of any CustomOrDefaultSetting fields on the
```
Good you added comments like this when code is not obvious without it

andrew-oc replied:
:thumbsup:

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/InputParameterEditor.cs:
```diff
@@ -30,6 +31,17 @@ public abstract class InputParameterEditor
+        /// changes value.</param>
+        public virtual void OnDrawVisualElements(VisualElement root, Action onChangedCallback)
+        {
+
```
Is this missing implementation intended? If it is, maybe its good with a inline comment explaining why its not relevant?

andrew-oc replied:
Oversight I think. Made it abstract.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/InputParameterEditor.cs:
```diff
@@ -200,6 +212,81 @@ internal struct CustomOrDefaultSetting
+	            // If previous value was an epsilon away from default value, it most likely means that value was set by our own code down in this method.
+	            // Revert it back to default to show a nice readable value in UI.
+	            // ReSharper disable once CompareOfFloatsByEqualityOperator
+	            if ((value - float.Epsilon) == m_DefaultInitializedValue)
```
Is it really intended to compare without epsilon here? That seems to indicate that defaultInitializedValue is 0 and in that case it might as well compare to a constant. For any other comparison this might go wrong?

andrew-oc replied:
Yep, this is really why we need tri-state values for this stuff, but as it was in the original code (and as far as I remember, implemented as a fix to a bug), I've left it as is for now. Ideally we'll be deleting large portions of this code in the future

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/InputParameterEditor.cs:
```diff
@@ -200,6 +212,81 @@ internal struct CustomOrDefaultSetting
+	            if (m_UseDefaultValue) return;
+
+	            // ReSharper disable once CompareOfFloatsByEqualityOperator
+	            if (evt.newValue == m_DefaultInitializedValue)
```
Maybe add braces here if having inline comments inside if statement to avoid future problems?

ekcoh replied:
Also same comment as before about float equality

ekcoh replied:
hmmm.... maybe its fine when I think of it because you already have capture round-off error in m_DefaultInitializedValue?

andrew-oc replied:
Added braces

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/InputParameterEditor.cs:
```diff
@@ -200,6 +212,81 @@ internal struct CustomOrDefaultSetting
+	            // ReSharper disable once CompareOfFloatsByEqualityOperator
+	            if (evt.newValue == m_DefaultInitializedValue)
+		            // If user sets a value that is equal to default initialized, change value slightly so it doesn't pass potential default checks.
+		            ////TODO: refactor all of this to use tri-state values instead, there is no obvious float value that we can use as default (well maybe NaN),
```
Agree, a tri-state is likely safer for this use-case

ekcoh replied:
Intended to be added as part of adding additional functionality? E.g. TODOs catches your eye while looking at the code but gets forgotten about whennot reviewing code, so maybe make sure we have this as part of a planned work task to be done?

andrew-oc replied:
This is why you need resharper :) 
![image](https://user-images.githubusercontent.com/78356434/163124580-def5a6d7-41c6-4646-b21c-97e441b0fc79.png)

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ViewBase.cs:
```diff
@@ -0,0 +1,291 @@
+
+		public virtual void RedrawUI(TViewState viewState)
+		{
+			
```
Intentionally empty?

andrew-oc replied:
I didn't want every view to have to implement RedrawUI, but it seems they all do now anyway, so made this one abstract.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ViewBase.cs:
```diff
@@ -0,0 +1,291 @@
+		/// </summary>
+		public virtual void DestroyView()
+		{
+
```
Intentionally empty?

ekcoh replied:
Would have expected it to call the generic version of it?

andrew-oc replied:
In this case, yes, since not all views need this. 

> Would have expected it to call the generic version of it?

It's actually the other way around. The generic version calls this one. I've renamed the generic version to DestroyChildViews now, which is a better name for what it does, but we do need to massage this bit of the design a little.

ekcoh reviewed (approved):
In general code looks very clean so great job on that part, but formatting is off in a number of places so run formatter on the code, some odd tabs etc observed to mitigate that. 
Liked that you extended existing editors with OnDrawVisualElements which provides side-by-side support which makes it easy to compare/follow while we need to support multiple UI systems.
I had a hard time understanding everything that is going on in viewstate parts without running the code (I did not as part of this review), but expecting to pick this up as part of actually working with this part of the code.

Approved (but contains some placeholders I have commented on), but expect that these are added as part of finalizing the work.

andrew-oc committed:
Added project context menu to open legacy editor from input asset

andrew-oc committed:
Added ability to assign bindings to control schemes

andrew-oc committed:
Formatting

andrew-oc committed:
Fix issue in CompositeBindingPropertoesView where parameter values would not be updated correctly

andrew-oc committed:
Review changes

andrew-oc committed:
Added missing UNITY_EDITOR ifdefs

andrew-oc committed:
Change dry run CI config to use 2022.2.0a10

andrew-oc committed:
Only show the UITK editor when the feature flag is set, and only on 2022.1

andrew-oc committed:
Some changes to get things compiling on versions of Unity less than 2022.1

andrew-oc committed:
More changes to fix build on 2019.4

andrew-oc committed:
Update package to 1.5

The InputParameterEditor class had OnDrawVisualElements public method added to it.

andrew-oc committed:
Fix incorrect UIToolkit uss reference

andrew-oc committed:
Revert "Change yamato config to build develop-v2 branch, but only build against trunk"

This reverts commit aaa54c18b25a156aab3badd85bc7bfc4813d36dc.

andrew-oc committed:
Revert "Update package to 1.5"

This reverts commit 8657d1e6521243395c58e9c0c2fecabd385e1979.

andrew-oc committed:
Moved some UITK editor specific code behind 2022.1 conditionals

The UITK asset editor is only usable on those versions of Unity

andrew-oc committed:
Merge branch 'develop' into project-settings-input-actions-editor

jfreire-unity commented:
@andrew-oc is this PR still relevant? Can it merged or can we close it?

jamesmcgill commented:
Still relevant. One of the current sprint goals is to get this merged.

jamesmcgill committed:
NEW: Add save and auto-save buttons

* NEW: Add save and auto-save buttons

* NEW: Recreate editor UI after domain reload

* Review changes: restore SerializeObject in InputActionEditorState

* Review changes: Move save logic out of the View

* Review changes: Move save logic into the Window

jamesmcgill committed:
NEW: UITK Editor separate action maps and actions views (#1661)

* Replaced action map drop down with ListView and bindings view with tree view showing actions and bindings

* Formatter

* Disabled code for renaming items in action maps and actions view

* Formatter

* Replaced action map drop down with ListView and bindings view with tree view showing actions and bindings

---------

Co-authored-by: Andrew O'Connor <andrew.oconnor@unity3d.com>

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1526) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1526) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/InputSystem?pullRequest=1526) it.</sub>

jamesmcgill committed:
fix merge error (again)

jamesmcgill committed:
fix formatting

jamesmcgill commented:
This work was merged in another PR: https://github.com/Unity-Technologies/InputSystem/pull/1668
